### PR TITLE
fix: updates validate_date check for publishing draft event

### LINKED ISSUE #6244: User can publish past events
**Describe the bug**
Past draft events can be published

**To Reproduce**
Create a draft event with today's date. Publish it tomorrow.

**Expected behavior**
Publishing past draft events should fail

### PR DESCRIPTION
**Pull request by shreyanshdwivedi**
Fixes #6244 

#### Short description of what this resolves:
User can publish past events.

#### Changes proposed in this pull request:
- restoring deleted event will have draft status if the date is behind current date and time
- while saving a draft, the date will not be checked
- while updating the status from draft to published will trigger validate_date and change the state to draft if date is behind current date and time

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

**Comment by shreyanshdwivedi** [comment]
@iamareebjamal @uds5501 @mrsaicharan1 please review

File: app/api/events.py [hunk 1]
```diff
@@ -492,7 +496,12 @@ def before_update_object(self, event, data, view_kwargs):
         :param view_kwargs:
         :return:
         """
-        if data.get('starts_at') != event.starts_at or data.get('ends_at') != event.ends_at:
+        # deleting event will not trigger validate_date
+        # change of event state from draft to published and also event restoring will trigger validate_date
+        if (data.get('starts_at') and data.get('starts_at') != event.starts_at) or \
+                (data.get('ends_at') and data.get('ends_at') != event.ends_at) or \
+                (data.get('state') == 'published' and event.state == 'draft') or \
+                (event.deleted_at and not data.get('deleted_at')):
```
Comment (iamareebjamal) [inline app/api/events.py]: Too much complexity. Most of these checks are present in validate_date. Why is there a need to add so many if else. Simply check if event.state == published or data.state == published

Comment (shreyanshdwivedi) [reply app/api/events.py]: I agree. Updated the checks

**Comment by iamareebjamal** [comment]
Verify other changes in PATCH don't affect this

**Comment by shreyanshdwivedi** [comment]
@iamareebjamal I think we need the previous code block I added to update_object -
```
    if data.get('starts_at') != event.starts_at or data.get('ends_at') != event.ends_at or \
                (event.state == "draft" and data.get('state') == "published") or \
                (event.deleted_at and not data.get('deleted_at')):
            validate_date(event, data)
```
- 1st line to check if there is any date change
- 2nd to check when we publish a drafted event
- 3rd to check when we restore event, so that we can update the status (if necessary)

**Comment by iamareebjamal** [comment]
OK, split into 3 separate booleans with comments so that it is understandanble

**Comment by shreyanshdwivedi** [comment]
@iamareebjamal added comments for better understanding

File: app/api/events.py [hunk 2]
```diff
@@ -492,8 +498,15 @@ def before_update_object(self, event, data, view_kwargs):
...
         """
+        # call validate_date if either event start_at or end_at is updated
         if data.get('starts_at') != event.starts_at or data.get('ends_at') != event.ends_at:
             validate_date(event, data)
+        # call validate_date if drafted event is being published
+        elif event.state == "draft" and data.get('state') == "published":
+            validate_date(event, data)
+        # call validate_date if event is being restored
+        elif event.deleted_at and not data.get('deleted_at'):
+            validate_date(event, data)
```
Comment (iamareebjamal) [inline app/api/events.py]: All 3 branches are same. Create 3 booleans and do or

File: app/api/events.py [hunk 3]
```diff
@@ -550,7 +556,12 @@ def before_update_object(self, event, data, view_kwargs):
         :param view_kwargs:
         :return:
         """
-        if data.get('starts_at') != event.starts_at or data.get('ends_at') != event.ends_at:
+        # 1 - call validate_date if either event start_at or end_at is updated
+        # 2 - call validate_date if drafted event is being published
+        # 3 - call validate_date if event is being restored
+        if (data.get('starts_at') != event.starts_at or data.get('ends_at') != event.ends_at) or \
+                (event.state == "draft" and data.get('state') == "published") or \
+                (event.deleted_at and not data.get('deleted_at')):
```
Comment (iamareebjamal) [inline app/api/events.py]: Still one big complex condition. Split in 3 booleans as mentioned previously. This is the third time now

Comment (shreyanshdwivedi) [reply app/api/events.py]: Maybe I'm not getting you.
Do you want me to have 3 different functions that return true or false and each function have one of the conditions ?

Comment (iamareebjamal) [reply app/api/events.py]: a = blah
b = blah
c =blah
a or b or c

File: app/api/events.py [hunk 4]
```diff
@@ -556,12 +556,14 @@ def before_update_object(self, event, data, view_kwargs):
...
-        # 3 - call validate_date if event is being restored
-        if (data.get('starts_at') != event.starts_at or data.get('ends_at') != event.ends_at) or \
-                (event.state == "draft" and data.get('state') == "published") or \
-                (event.deleted_at and not data.get('deleted_at')):
+        # is either event start_at or end_at is updated
+        is_date_updated = (data.get('starts_at') != event.starts_at or
+                           data.get('ends_at') != event.ends_at)
+        # is drafted event is being published
+        is_draft_published = (event.state == "draft" and data.get('state') == "published")
+        # is deleted event is being restored
```
Comment (iamareebjamal) [inline app/api/events.py]: Remove comments now, variable names are seld explanatory

Comment (shreyanshdwivedi) [reply app/api/events.py]: Made the changes. Please have a look

**Pull request merged**